File: src/DashSpec.Core/Parsing/CardDiagramOverrideParser.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal static class CardDiagramOverrideParser
{
    internal sealed record DiagramDelta(
        DiagramDefinition? Diagram,
        LegendDefinition? Legend,
        PresentationBlock? Presentation,
        SeriesTransformBlock? SeriesTransform);

    public static void ParseOverridesBlock(
        TokenReader reader,
        string cardId,
        bool plural,
        ref DiagramDefinition? diagram,
        ref LegendDefinition? legend,
        ref PresentationBlock? presentation,
        ref SeriesTransformBlock? seriesTransform)
    {
        if (plural)
        {
            BlockSyntax.BeginBlock(reader);
            reader.SkipNewlines();
            while (!BlockSyntax.IsBlockEnd(reader, "overrides") && !reader.IsEof)
            {
                reader.SkipNewlines();
                if (BlockSyntax.IsBlockEnd(reader, "overrides"))
                {
                    break;
                }

                if (!reader.TryKeyword("for"))
                {
                    throw reader.Unexpected("for <diagram_id>");
                }

                var targetId = reader.ReadIdent();
                var delta = ParseOverrideBody(reader, targetId, cardId, endKind: "for");
                ApplyDelta(ref diagram, ref legend, ref presentation, ref seriesTransform, delta, targetId, cardId);
            }

            BlockSyntax.ExpectBlockEnd(reader, "overrides");
            return;
        }

        if (!reader.TryKeyword("for"))
        {
            throw reader.Unexpected("for <diagram_id>");
        }

        var diagramId = reader.ReadIdent();
        var single = ParseOverrideBody(reader, diagramId, cardId, endKind: "override");
        ApplyDelta(ref diagram, ref legend, ref presentation, ref seriesTransform, single, diagramId, cardId);
    }

    public static DiagramDelta? TryParseDiagramInlineBody(
        TokenReader reader,
        string diagramId,
        string cardId,
        string parentEndKind)
    {
        reader.SkipNewlines();
        if (BlockSyntax.IsBlockEnd(reader, parentEndKind) || reader.IsEof)
        {
            return null;
        }

        if (reader.TryPeekIdent(out var next) &&
            (string.Equals(next, "diagram", StringComparison.OrdinalIgnoreCase) ||
             string.Equals(next, "legend", StringComparison.OrdinalIgnoreCase) ||
             string.Equals(next, "datasource", StringComparison.OrdinalIgnoreCase) ||
             string.Equals(next, "bind", StringComparison.OrdinalIgnoreCase) ||
             string.Equals(next, "chrome", StringComparison.OrdinalIgnoreCase) ||
             string.Equals(next, "when", StringComparison.OrdinalIgnoreCase) ||
             string.Equals(next, "layout", StringComparison.OrdinalIgnoreCase) ||
             string.Equals(next, "override", StringComparison.OrdinalIgnoreCase) ||
             string.Equals(next, "data", StringComparison.OrdinalIgnoreCase) ||
             string.Equals(next, "view", StringComparison.OrdinalIgnoreCase)))
        {
            return null;
        }

        return ParseOverrideBody(reader, diagramId, cardId, endKind: "diagram", endId: diagramId);
    }

    private static DiagramDelta ParseOverrideBody(
        TokenReader reader,
        string diagramId,
        string cardId,
        string endKind,
        string? endId = null)
    {
        BlockSyntax.BeginBlock(reader);
        reader.SkipNewlines();

        var props = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        LegendDefinition? legend = null;
        PresentationBlock? presentation = null;
        SeriesTransformBlock? seriesTransform = null;

        while (!BlockSyntax.IsBlockEnd(reader, endKind, endId) && !reader.IsEof)
        {
            reader.SkipNewlines();
            if (BlockSyntax.IsBlockEnd(reader, endKind, endId))
            {
                break;
            }

            if (reader.TryKeyword("series"))
            {
                seriesTransform = ParseSeriesOverride(reader, cardId, diagramId);
                continue;
            }

            if (reader.TryKeyword("legend"))
            {
                legend = ParseLegendOverride(reader);
                continue;
            }

            if (reader.TryKeyword("presentation"))
            {
                presentation = ParsePresentationOverride(reader);
                continue;
            }

            if (reader.TryPeekIdent(out var key) && reader.RawKind is TokenKind.Ident)
            {
                _ = reader.ReadIdent();
                reader.Expect(TokenKind.Eq);
                props[key] = reader.ReadScalarValue();
                reader.SkipNewlines();
                continue;
            }

            throw reader.Unexpected();
        }

        BlockSyntax.ExpectBlockEnd(reader, endKind, endId);

        var diagram = props.Count > 0
            ? new DiagramDefinition(string.Empty, props, diagramId)
            : null;

        return new DiagramDelta(diagram, legend, presentation, seriesTransform);
    }

    private static SeriesTransformBlock ParseSeriesOverride(TokenReader reader, string cardId, string diagramId)
    {
        if (reader.TryKeyword("max"))
        {
            reader.Expect(TokenKind.Eq);
            if (!int.TryParse(reader.ReadScalarValue(), out var max) || max <= 0)
            {
                throw new DashSpecParseException(
                    $"Card '{cardId}': override for '{diagramId}' series max must be a positive integer.");
            }

            return new SeriesTransformBlock(null, max, null);
        }

        if (reader.IsOnNewline())
        {
            reader.SkipNewlines();
            var props = PropertyBlockParser.Parse(reader, PropertySchemas.SeriesTransform, "series", endKind: "series");
            props.TryGetValue("use", out var usePreset);
            int? maxValue = null;
            if (props.TryGetValue("max", out var rawMax) &&
                int.TryParse(rawMax, out var parsedMax) &&
                parsedMax > 0)
            {
                maxValue = parsedMax;
            }

            props.TryGetValue("other", out var otherLabel);
            return new SeriesTransformBlock(usePreset, maxValue, otherLabel);
        }

        throw reader.Unexpected("max or multiline series block");
    }

    private static LegendDefinition ParseLegendOverride(TokenReader reader)
    {
        var props = PropertyBlockParser.Parse(reader, PropertySchemas.Legend, "legend", endKind: "legend");
        return new LegendDefinition(
            props.GetValueOrDefault("min"),
            props.GetValueOrDefault("max"),
            props.GetValueOrDefault("title"));
    }

    private static PresentationBlock ParsePresentationOverride(TokenReader reader)
    {
        var props = PropertyBlockParser.Parse(reader, PropertySchemas.Presentation, "presentation", endKind: "presentation");
        props.TryGetValue("use", out var usePreset);
        var inline = props
            .Where(x => !string.Equals(x.Key, "use", StringComparison.OrdinalIgnoreCase))
            .ToDictionary(x => x.Key, x => x.Value, StringComparer.OrdinalIgnoreCase);
        return new PresentationBlock(usePreset, inline);
    }

    private static void ApplyDelta(
        ref DiagramDefinition? diagram,
        ref LegendDefinition? legend,
        ref PresentationBlock? presentation,
        ref SeriesTransformBlock? seriesTransform,
        DiagramDelta delta,
        string targetId,
        string cardId)
    {
        if (diagram is not null &&
            !string.IsNullOrWhiteSpace(diagram.UsePreset) &&
            !string.Equals(diagram.UsePreset, targetId, StringComparison.OrdinalIgnoreCase))
        {
            throw new DashSpecParseException(
                $"Card '{cardId}': override for '{targetId}' does not match view diagram '{diagram.UsePreset}'.");
        }

        if (delta.Diagram is not null)
        {
            diagram = diagram is null
                ? delta.Diagram
                : SpecIncludeResolver.Merge(
                    new SpecIncludeFragment(delta.Diagram, null, null),
                    new SpecIncludeFragment(diagram, null, null)).Diagram;
        }

        if (delta.Legend is not null)
        {
            legend = delta.Legend;
        }

        if (delta.Presentation is not null)
        {
            presentation = presentation is null
                ? delta.Presentation
                : SpecIncludeResolver.Merge(
                    new SpecIncludeFragment(null, delta.Presentation, null),
                    new SpecIncludeFragment(null, presentation, null)).Presentation;
        }

        if (delta.SeriesTransform is not null)
        {
            seriesTransform = seriesTransform is null
                ? delta.SeriesTransform
                : SpecIncludeResolver.Merge(
                    new SpecIncludeFragment(null, null, delta.SeriesTransform),
                    new SpecIncludeFragment(null, null, seriesTransform)).SeriesTransform;
        }
    }
}

File: src/DashSpec.Modeling.Parse/Card/CardDiagramOverrideParser.fs
namespace DashSpec.Modeling.Parse.Card

open System
open System.Collections.Generic
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Diagram
open DashSpec.Modeling.Parse.Include
open DashSpec.Modeling.Parse.Lexing
open DashSpec.Modeling.Parse.Presentation
open DashSpec.Modeling.Parse.Transform

module CardDiagramOverrideParser =

    type DiagramDelta =
        { Diagram: DiagramDefinition option
          Legend: LegendDefinition option
          Presentation: PresentationBlock option
          SeriesTransform: SeriesTransformBlock option }

    let private toPresentationBlock (props: IDictionary<string, string>) =
        let usePreset =
            match props.TryGetValue "use" with
            | true, value -> Some value
            | false, _ -> None
        let inlineProps =
            let map = Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            for kv in props do
                if not (kv.Key.Equals("use", StringComparison.OrdinalIgnoreCase)) then
                    map.[kv.Key] <- kv.Value
            map :> IReadOnlyDictionary<_, _>
        { UsePreset = usePreset; Properties = inlineProps }

    let private propValue (props: IDictionary<string, string>) key =
        match props.TryGetValue key with
        | true, value -> Some value
        | false, _ -> None

    let private parseLegendOverride (reader: TokenReader) =
        let props =
            PropertyBlockParser.parseWithEndKind reader PropertySchemas.legend "legend" "legend" false false
        { MinLabel = propValue props "min"
          MaxLabel = propValue props "max"
          Title = propValue props "title" }

    let private parsePresentationOverride (reader: TokenReader) =
        let props =
            PropertyBlockParser.parseWithEndKind reader PropertySchemas.presentation "presentation" "presentation" false false
        toPresentationBlock props

    let private parseSeriesOverride (reader: TokenReader) (cardId: string) (diagramId: string) =
        if reader.TryKeyword "max" then
            reader.Expect TokenKind.Eq
            let mutable max = 0
            if not (Int32.TryParse(reader.ReadScalarValue(), &max) && max > 0) then
                raise (DashSpecParseException($"Card '{cardId}': override for '{diagramId}' series max must be a positive integer."))
            { UsePreset = None; Max = Some max; OtherLabel = None }
        elif reader.IsOnNewline() then
            reader.SkipNewlines()
            let props =
                PropertyBlockParser.parseWithEndKind reader PropertySchemas.seriesTransform "series" "series" false false
            let usePreset = propValue props "use"
            let max =
                match propValue props "max" with
                | Some raw ->
                    let mutable parsed = 0
                    if Int32.TryParse(raw, &parsed) && parsed > 0 then Some parsed else None
                | None -> None
            let other = propValue props "other"
            { UsePreset = usePreset; Max = max; OtherLabel = other }
        else
            raise (reader.Unexpected "max or multiline series block")

    let private parseOverrideBody (reader: TokenReader) (diagramId: string) (cardId: string) (endKind: string) (endId: string option) =
        BlockSyntax.beginBlock reader
        reader.SkipNewlines()
        let props = Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        let mutable legend: LegendDefinition option = None
        let mutable presentation: PresentationBlock option = None
        let mutable seriesTransform: SeriesTransformBlock option = None

        while not (BlockSyntax.isBlockEnd reader endKind endId) && not reader.IsEof do
            reader.SkipNewlines()
            if BlockSyntax.isBlockEnd reader endKind endId then ()
            elif reader.TryKeyword "series" then
                seriesTransform <- Some(parseSeriesOverride reader cardId diagramId)
            elif reader.TryKeyword "legend" then
                legend <- Some(parseLegendOverride reader)
            elif reader.TryKeyword "presentation" then
                presentation <- Some(parsePresentationOverride reader)
            elif reader.TryPeekIdent().IsSome && reader.RawKind = TokenKind.Ident then
                let key = reader.ReadIdent()
                reader.Expect TokenKind.Eq
                props.[key] <- reader.ReadScalarValue()
                reader.SkipNewlines()
            else
                raise (reader.Unexpected())

        BlockSyntax.expectBlockEnd reader endKind endId

        let diagram =
            if props.Count > 0 then
                Some { Kind = ""; Properties = props :> IReadOnlyDictionary<_, _>; UsePreset = Some diagramId }
            else
                None

        { Diagram = diagram; Legend = legend; Presentation = presentation; SeriesTransform = seriesTransform }

    let private applyDelta
        (diagram: DiagramDefinition option ref)
        (legend: LegendDefinition option ref)
        (presentation: PresentationBlock option ref)
        (seriesTransform: SeriesTransformBlock option ref)
        (delta: DiagramDelta)
        (targetId: string)
        (cardId: string)
        =
        match diagram.Value with
        | Some d when d.UsePreset.IsSome && not (String.Equals(d.UsePreset.Value, targetId, StringComparison.OrdinalIgnoreCase)) ->
            raise (DashSpecParseException($"Card '{cardId}': override for '{targetId}' does not match view diagram '{d.UsePreset.Value}'."))
        | _ -> ()

        if delta.Diagram.IsSome then
            diagram.Value <-
                match diagram.Value with
                | None -> delta.Diagram
                | Some current ->
                    (SpecIncludeFragmentResolver.merge
                        { SpecIncludeFragmentResolver.emptyFragment with Diagram = Some current }
                        { SpecIncludeFragmentResolver.emptyFragment with Diagram = delta.Diagram }).Diagram

        if delta.Legend.IsSome then legend.Value <- delta.Legend
        if delta.Presentation.IsSome then
            presentation.Value <-
                match presentation.Value with
                | None -> delta.Presentation
                | Some current ->
                    (SpecIncludeFragmentResolver.merge
                        { SpecIncludeFragmentResolver.emptyFragment with Presentation = Some current }
                        { SpecIncludeFragmentResolver.emptyFragment with Presentation = delta.Presentation }).Presentation
        if delta.SeriesTransform.IsSome then
            seriesTransform.Value <-
                match seriesTransform.Value with
                | None -> delta.SeriesTransform
                | Some current ->
                    (SpecIncludeFragmentResolver.merge
                        { SpecIncludeFragmentResolver.emptyFragment with SeriesTransform = Some current }
                        { SpecIncludeFragmentResolver.emptyFragment with SeriesTransform = delta.SeriesTransform }).SeriesTransform

    let parseOverridesBlock
        (reader: TokenReader)
        (cardId: string)
        (plural: bool)
        (diagram: DiagramDefinition option ref)
        (legend: LegendDefinition option ref)
        (presentation: PresentationBlock option ref)
        (seriesTransform: SeriesTransformBlock option ref)
        =
        if plural then
            BlockSyntax.beginBlock reader
            reader.SkipNewlines()
            while not (BlockSyntax.isBlockEnd reader "overrides" None) && not reader.IsEof do
                reader.SkipNewlines()
                if BlockSyntax.isBlockEnd reader "overrides" None then ()
                elif not (reader.TryKeyword "for") then
                    raise (reader.Unexpected "for <diagram_id>")
                else
                    let targetId = reader.ReadIdent()
                    let delta = parseOverrideBody reader targetId cardId "for" None
                    applyDelta diagram legend presentation seriesTransform delta targetId cardId
            BlockSyntax.expectBlockEnd reader "overrides" None
        else
            if not (reader.TryKeyword "for") then raise (reader.Unexpected "for <diagram_id>")
            let diagramId = reader.ReadIdent()
            let delta = parseOverrideBody reader diagramId cardId "override" None
            applyDelta diagram legend presentation seriesTransform delta diagramId cardId

    let tryParseDiagramInlineBody (reader: TokenReader) (diagramId: string) (cardId: string) (parentEndKind: string) =
        reader.SkipNewlines()
        if BlockSyntax.isBlockEnd reader parentEndKind None || reader.IsEof then None
        elif
            match reader.TryPeekIdent() with
            | Some next ->
                String.Equals(next, "diagram", StringComparison.OrdinalIgnoreCase)
                || String.Equals(next, "legend", StringComparison.OrdinalIgnoreCase)
                || String.Equals(next, "datasource", StringComparison.OrdinalIgnoreCase)
                || String.Equals(next, "bind", StringComparison.OrdinalIgnoreCase)
                || String.Equals(next, "chrome", StringComparison.OrdinalIgnoreCase)
                || String.Equals(next, "when", StringComparison.OrdinalIgnoreCase)
                || String.Equals(next, "layout", StringComparison.OrdinalIgnoreCase)
                || String.Equals(next, "override", StringComparison.OrdinalIgnoreCase)
                || String.Equals(next, "data", StringComparison.OrdinalIgnoreCase)
                || String.Equals(next, "view", StringComparison.OrdinalIgnoreCase)
            | None -> false
        then
            None
        else
            Some(parseOverrideBody reader diagramId cardId "diagram" (Some diagramId))
